spec/core/ics-003-connection-semantics/UPGRADES.md

```diff
@@ -0,0 +1,495 @@
+    } else {
+        // abort transaction if current connection is not in INIT or OPEN
+        abortTransactionUnless(false)
+    }
+    
+    // either timeout height or timestamp must be non-zero
+    // if the upgrade feature is implemented on the TRY chain, then a relayer may submit a TRY transaction after the timeout.
+    // this will restore the connection on the executing chain and allow counterparty to use the CancelUpgradeMsg to restore their connection.
+    if timeoutHeight == 0 && timeoutTimestamp == 0 {
+        restoreConnection()
+        return
+    }
+    upgradeTimeout = UpgradeTimeout{
+        timeoutHeight: timeoutHeight,
+        timeoutTimestamp: timeoutTimestamp,
+    }
+
+    // counterparty-specified timeout must not have exceeded
+    if (currentHeight() > timeoutHeight && timeoutHeight != 0) ||
+        (currentTimestamp() > timeoutTimestamp && timeoutTimestamp != 0) {
+        restoreConnection()
+        return
+    }
+
+    // verify chosen versions are compatible
+    versionsIntersection = intersection(counterpartyConnection.version, proposedUpgradeConnection.version)
+    version = pickVersion(versionsIntersection) // aborts transaction if there is no intersection
+
+    // both connection ends must be mutually compatible.
+    // this function has been left unspecified since it will depend on the specific structure of the new connection.
+    // It is the responsibility of implementations to make sure that verification that the proposed new connections
+    // on either side are correctly constructed according to the new version selected.
+    if !IsCompatible(counterpartyConnection, proposedUpgradeConnection) {
+        restoreConnection()
+        return
+    }
+
+    provableStore.set(connectionPath(identifier), proposedUpgradeConnection)
+}
+```
+
+NOTE: It is up to individual implementations how they will provide access-control to the `ConnUpgradeTry` function. E.g. chain governance, permissioned actor, DAO, etc. A chain may decide to have permissioned **or** permissionless `UpgradeTry`. In the permissioned case, both chains must explicitly consent to the upgrade, in the permissionless case; one chain initiates the upgrade and the other chain agrees to the upgrade by default. In the permissionless case, a relayer may submit the `ConnUpgradeTry` datagram.
+
+
+```typescript
+function connUpgradeAck(
+    identifier: Identifier,
+    counterpartyConnection: ConnectionEnd,
+    proofConnection: CommitmentProof,
+    proofHeight: Height
+) {
+    // current connection is in UPGRADE_INIT or UPGRADE_TRY (crossing hellos)
+    currentConnection = provableStore.get(connectionPath(identifier))
+    abortTransactionUnless(currentConnection.state == UPGRADE_INIT || currentConnection.state == UPGRADE_TRY)
+
+    // counterparty must be in TRY state
+    if counterpartyConnection.State != UPGRADE_TRY {
+        restoreConnection()
+        return
+    }
```

Comment:
we have to be careful to not use relayer provided information unless verified (especially if we are restoring/cancelling an upgrade)